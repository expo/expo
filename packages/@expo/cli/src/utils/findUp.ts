import path from 'path';
import resolveFrom from 'resolve-from';

import { CommandError } from '../utils/errors';

/** Look up directories until one with a `package.json` can be found, assert if none can be found. */
export function findUpProjectRootOrAssert(cwd: string): string {
  const projectRoot = findUpProjectRoot(cwd);
  if (!projectRoot) {
    throw new CommandError(`Project root directory not found (working directory: ${cwd})`);
  }
  return projectRoot;
}

function findUpProjectRoot(cwd: string): string | null {
  if (['.', path.sep].includes(cwd)) return null;

  const found = resolveFrom.silent(cwd, './package.json');
  if (found) {
    return path.dirname(found);
  }
  return findUpProjectRoot(path.dirname(cwd));
}
