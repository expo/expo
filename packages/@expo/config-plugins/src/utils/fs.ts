import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}

/** Remove a single file (not directory). Returns `true` if a file was actually deleted. */
export function removeFile(filePath: string): boolean {
  try {
    fs.unlinkSync(filePath);
    return true;
  } catch (error: any) {
    // Skip if the remove did nothing.
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
