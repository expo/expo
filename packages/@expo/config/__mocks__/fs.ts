module.exports = require('memfs').fs;

// NOTE(cedric): workaround to also mock `node:fs`
jest.mock('node:fs', () => require('fs'));
