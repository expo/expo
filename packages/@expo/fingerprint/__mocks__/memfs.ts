const memfs = jest.requireActual('memfs');
const path = require('path');

/**
 * Further mock memfs `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = memfs.fs.promises.mkdtemp;
memfs.fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  memfs.fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = memfs;
