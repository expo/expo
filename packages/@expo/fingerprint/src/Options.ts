import os from 'os';

import type { NormalizedOptions, Options } from './Fingerprint.types';

export function normalizeOptions(options?: Options): NormalizedOptions {
  return {
    ...options,
    platforms: options?.platforms ?? ['android', 'ios'],
    concurrentIoLimit: options?.concurrentIoLimit ?? os.cpus().length,
    hashAlgorithm: options?.hashAlgorithm ?? 'sha1',
    dirExcludes: options?.dirExcludes ?? [
      '**/android/build',
      '**/android/app/build',
      '**/android/app/.cxx',
      'ios/Pods',
    ],
  };
}
