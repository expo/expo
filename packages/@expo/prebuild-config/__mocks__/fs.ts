const fs = require('memfs').fs;

// Add fs.promises.cp polyfill for memfs since it doesn't support it
fs.promises.cp = async (src: string, dest: string, options?: { recursive?: boolean }) => {
  const path = require('path');

  const srcStat = await fs.promises.stat(src);

  if (srcStat.isDirectory()) {
    await fs.promises.mkdir(dest, { recursive: true });
    const entries = await fs.promises.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await fs.promises.cp(srcPath, destPath, options);
      } else {
        await fs.promises.copyFile(srcPath, destPath);
      }
    }
  } else {
    await fs.promises.copyFile(src, dest);
  }
};

module.exports = fs;

jest.mock('node:fs', () => require('memfs').fs);
