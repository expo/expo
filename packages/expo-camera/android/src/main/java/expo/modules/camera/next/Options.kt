package expo.modules.camera.next

import expo.modules.camera.next.records.VideoQuality
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

data class PictureOptions(
  @Field val quality: Double = 1.0,
  @Field val base64: Boolean = false,
  @Field val exif: Boolean = false,
  @Field val additionalExif: Map<String, Any>? = null,
  @Field val skipProcessing: Boolean = false,
  @Field val fastMode: Boolean = false,
  @Field val id: Int? = null,
  @Field val maxDownsampling: Int = 1
) : Record

data class RecordingOptions(
  @Field val maxDuration: Int = 0,
  @Field val maxFileSize: Int = 0,
  @Field val quality: VideoQuality?
) : Record
