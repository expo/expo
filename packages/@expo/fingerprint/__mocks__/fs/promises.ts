import { fs } from 'memfs';
import path from 'path';

/**
 * Further mock `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = fs.promises.mkdtemp;
fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = fs.promises;
