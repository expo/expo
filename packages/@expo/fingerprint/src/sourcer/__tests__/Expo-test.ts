import spawnAsync from '@expo/spawn-async';
import findUp from 'find-up';
import fs from 'fs';
import { vol, fs as volFS } from 'memfs';
import path from 'path';
import resolveFrom from 'resolve-from';

import { normalizeOptions } from '../../Options';
import {
  getEasBuildSourcesAsync,
  getExpoAutolinkingAndroidSourcesAsync,
  getExpoAutolinkingIosSourcesAsync,
  getExpoConfigSourcesAsync,
} from '../Expo';

jest.mock('@expo/spawn-async');
jest.mock('find-up');
jest.mock('fs/promises');
jest.mock('resolve-from');
jest.mock('/app/package.json', () => {}, { virtual: true });

describe(getEasBuildSourcesAsync, () => {
  afterEach(() => {
    vol.reset();
  });

  it('should contains `eas.json` file', async () => {
    vol.fromJSON(require('./fixtures/ExpoManaged47Project.json'));
    vol.writeFileSync(
      '/app/eas.json',
      `
{
  "cli": {
    "version": ">= 2.6.0"
  },
  "build": {
    "development": {
      "distribution": "internal",
      "android": {
        "gradleCommand": ":app:assembleDebug"
      },
      "ios": {
        "buildConfiguration": "Debug"
      }
    },
    "preview": {
      "distribution": "internal"
    },
    "production": {}
  },
  "submit": {
    "production": {}
  }
}`
    );

    const sources = await getEasBuildSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'file',
        filePath: 'eas.json',
      })
    );
  });
});

describe('getExpoAutolinkingSourcesAsync', () => {
  beforeEach(() => {
    const mockSpawnAsync = spawnAsync as jest.MockedFunction<typeof spawnAsync>;
    const fixtureAndroid = fs.readFileSync(
      path.join(__dirname, 'fixtures', 'ExpoAutolinkingAndroid.json'),
      'utf8'
    );
    const fixtureIos = fs.readFileSync(
      path.join(__dirname, 'fixtures', 'ExpoAutolinkingIos.json'),
      'utf8'
    );
    mockSpawnAsync.mockResolvedValueOnce({
      stdout: fixtureAndroid,
      stderr: '',
      status: 0,
      signal: null,
      output: [fixtureAndroid, ''],
    });
    mockSpawnAsync.mockResolvedValueOnce({
      stdout: fixtureIos,
      stderr: '',
      status: 0,
      signal: null,
      output: [fixtureIos, ''],
    });
  });

  afterEach(() => {
    vol.reset();
  });

  it('should contain expo autolinking projects', async () => {
    let sources = await getExpoAutolinkingAndroidSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'dir',
        filePath: 'node_modules/expo-modules-core/android',
      })
    );
    expect(sources).toMatchSnapshot();

    sources = await getExpoAutolinkingIosSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({ type: 'dir', filePath: 'node_modules/expo-modules-core' })
    );
    expect(sources).toMatchSnapshot();
  });

  it('should not containt absolute path in contents', async () => {
    let sources = await getExpoAutolinkingAndroidSourcesAsync('/app', normalizeOptions());
    for (const source of sources) {
      if (source.type === 'contents') {
        expect(source.contents.indexOf('/app/')).toBe(-1);
      }
    }

    sources = await getExpoAutolinkingIosSourcesAsync('/app', normalizeOptions());
    for (const source of sources) {
      if (source.type === 'contents') {
        expect(source.contents.indexOf('/app/')).toBe(-1);
      }
    }
  });
});

describe(getExpoConfigSourcesAsync, () => {
  beforeAll(() => {
    jest.doMock('fs', () => volFS);
  });

  afterEach(() => {
    vol.reset();
  });

  it('should contain app.json', async () => {
    vol.fromJSON(require('./fixtures/ExpoManaged47Project.json'));
    const sources = await getExpoConfigSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'file',
        filePath: 'app.json',
      })
    );
  });

  it('should contain external icon file in app.json', async () => {
    vol.fromJSON(require('./fixtures/ExpoManaged47Project.json'));
    vol.mkdirSync('/app/assets');
    vol.writeFileSync('/app/assets/icon.png', 'PNG data');
    const sources = await getExpoConfigSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'file',
        filePath: './assets/icon.png',
      })
    );
  });
});

describe(`getExpoConfigSourcesAsync - config-plugins`, () => {
  let baseAppJson: { expo: any };

  function setupThirdPartyPlugin() {
    vol.mkdirSync('/app/node_modules/third-party', { recursive: true });

    // package.json
    vol.writeFileSync('/app/node_modules/third-party/package.json', '{}');
    const mockFindUpSync = findUp.sync as jest.MockedFunction<typeof findUp.sync>;
    mockFindUpSync.mockReturnValue('/app/node_modules/third-party/package.json');

    // entry file
    const withNoopPlugin = (config: any) => config;
    jest.mock('/app/node_modules/third-party/index.js', () => ({ default: withNoopPlugin }), {
      virtual: true,
    });
    const mockResolveFrom = resolveFrom.silent as jest.MockedFunction<typeof resolveFrom.silent>;
    mockResolveFrom.mockReturnValue('/app/node_modules/third-party/index.js');
  }

  beforeEach(() => {
    jest.doMock('fs', () => volFS);
    vol.fromJSON(require('./fixtures/ExpoManaged47Project.json'));
    baseAppJson = JSON.parse(vol.readFileSync('/app/app.json', 'utf8').toString());
  });

  afterEach(() => {
    vol.reset();
    jest.resetAllMocks();
  });

  it('should contain external config-plugin dir', async () => {
    setupThirdPartyPlugin();

    vol.writeFileSync(
      '/app/app.json',
      JSON.stringify({
        ...baseAppJson,
        expo: {
          ...baseAppJson.expo,
          plugins: ['third-party'],
        },
      })
    );
    const sources = await getExpoConfigSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'dir',
        filePath: 'node_modules/third-party',
      })
    );
  });

  it('should contain external config-plugin dir from plugin with parameters', async () => {
    setupThirdPartyPlugin();

    vol.writeFileSync(
      '/app/app.json',
      JSON.stringify({
        ...baseAppJson,
        expo: {
          ...baseAppJson.expo,
          plugins: [['third-party', { parameter: 'foo' }]],
        },
      })
    );
    const sources = await getExpoConfigSourcesAsync('/app', normalizeOptions());
    expect(sources).toContainEqual(
      expect.objectContaining({
        type: 'dir',
        filePath: 'node_modules/third-party',
      })
    );
  });

  it('should not contain external config-plugin dir from raw function plugins', async () => {
    vol.writeFileSync(
      '/app/app.config.js',
      `\
export default ({ config }) => {
  return config;
};`
    );
    const sources = await getExpoConfigSourcesAsync('/app', normalizeOptions());

    vol.writeFileSync(
      '/app/app.config.js',
      `\
export default ({ config }) => {
  config.plugins ||= [];
  const withNoopPlugin = (config: any) => config;
  config.plugins.push(withNoopPlugin);
  return config;
};`
    );
    const sources2 = await getExpoConfigSourcesAsync('/app', normalizeOptions());

    expect(sources).toEqual(sources2);
  });
});
