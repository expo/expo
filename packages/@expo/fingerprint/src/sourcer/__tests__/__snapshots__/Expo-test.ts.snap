// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExpoAutolinkingSourcesAsync should contain expo autolinking projects 1`] = `
[
  {
    "filePath": "node_modules/expo/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-application/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-constants/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-error-recovery/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-file-system/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-font/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-keep-awake/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-modules-core/android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "dir",
  },
  {
    "contents": "{"modules":[{"packageName":"expo","packageVersion":"47.0.3","projects":[{"name":"expo","sourceDir":"node_modules/expo/android"}],"modules":[]},{"packageName":"expo-application","packageVersion":"5.0.1","projects":[{"name":"expo-application","sourceDir":"node_modules/expo-application/android"}],"modules":[]},{"packageName":"expo-constants","packageVersion":"14.0.2","projects":[{"name":"expo-constants","sourceDir":"node_modules/expo-constants/android"}],"modules":[]},{"packageName":"expo-error-recovery","packageVersion":"4.0.1","projects":[{"name":"expo-error-recovery","sourceDir":"node_modules/expo-error-recovery/android"}],"modules":[]},{"packageName":"expo-file-system","packageVersion":"15.1.1","projects":[{"name":"expo-file-system","sourceDir":"node_modules/expo-file-system/android"}],"modules":[]},{"packageName":"expo-font","packageVersion":"11.0.1","projects":[{"name":"expo-font","sourceDir":"node_modules/expo-font/android"}],"modules":[]},{"packageName":"expo-keep-awake","packageVersion":"11.0.1","projects":[{"name":"expo-keep-awake","sourceDir":"node_modules/expo-keep-awake/android"}],"modules":[]},{"packageName":"expo-modules-core","packageVersion":"1.0.2","projects":[{"name":"expo-modules-core","sourceDir":"node_modules/expo-modules-core/android"}],"modules":[]}]}",
    "id": "expoAutolinkingConfig:android",
    "reasons": [
      "expoAutolinkingAndroid",
    ],
    "type": "contents",
  },
]
`;

exports[`getExpoAutolinkingSourcesAsync should contain expo autolinking projects 2`] = `
[
  {
    "filePath": "node_modules/expo",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-application/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-constants/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-error-recovery/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-file-system/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-font/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-keep-awake/ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "filePath": "node_modules/expo-modules-core",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "dir",
  },
  {
    "contents": "{"modules":[{"packageName":"expo","packageVersion":"47.0.3","pods":[{"podName":"Expo","podspecDir":"node_modules/expo"}],"swiftModuleNames":["Expo"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-application","packageVersion":"5.0.1","pods":[{"podName":"EXApplication","podspecDir":"node_modules/expo-application/ios"}],"swiftModuleNames":["EXApplication"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-constants","packageVersion":"14.0.2","pods":[{"podName":"EXConstants","podspecDir":"node_modules/expo-constants/ios"}],"swiftModuleNames":["EXConstants"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-error-recovery","packageVersion":"4.0.1","pods":[{"podName":"EXErrorRecovery","podspecDir":"node_modules/expo-error-recovery/ios"}],"swiftModuleNames":["EXErrorRecovery"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-file-system","packageVersion":"15.1.1","pods":[{"podName":"EXFileSystem","podspecDir":"node_modules/expo-file-system/ios"}],"swiftModuleNames":["EXFileSystem"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-font","packageVersion":"11.0.1","pods":[{"podName":"EXFont","podspecDir":"node_modules/expo-font/ios"}],"swiftModuleNames":["EXFont"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-keep-awake","packageVersion":"11.0.1","pods":[{"podName":"ExpoKeepAwake","podspecDir":"node_modules/expo-keep-awake/ios"}],"swiftModuleNames":["ExpoKeepAwake"],"modules":["KeepAwakeModule"],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false},{"packageName":"expo-modules-core","packageVersion":"1.0.2","pods":[{"podName":"ExpoModulesCore","podspecDir":"node_modules/expo-modules-core"}],"swiftModuleNames":["ExpoModulesCore"],"modules":[],"appDelegateSubscribers":[],"reactDelegateHandlers":[],"debugOnly":false}]}",
    "id": "expoAutolinkingConfig:ios",
    "reasons": [
      "expoAutolinkingIos",
    ],
    "type": "contents",
  },
]
`;
